packages/plugin-dts/src/utils.ts

```diff
@@ -135,16 +164,178 @@ export async function addBannerAndFooter(
   }
 
   if (code.hasChanged()) {
-    await fsP.writeFile(file, code.toString());
+    await fsP.writeFile(dtsFile, code.toString());
+  }
+}
+
+export async function redirectDtsImports(
+  dtsFile: string,
+  dtsExtension: string,
+  tsconfigPath: string,
+  outDir: string,
+  rootDir: string,
+  redirect?: DtsRedirect,
+): Promise<void> {
+  if (!redirect || (!redirect.path && !redirect.extension)) {
+    return;
+  }
+
+  const extensions = dtsExtension
+    .replace(/\.d\.ts$/, '.js')
+    .replace(/\.d\.cts$/, '.cjs')
+    .replace(/\.d\.mts$/, '.mjs');
+  const content = await fsP.readFile(dtsFile, 'utf-8');
+  const code = new MagicString(content);
+  const sgNode = (await parseAsync('typescript', content)).root();
+  const matcher: NapiConfig = {
+    rule: {
+      kind: 'string_fragment',
+      any: [
+        {
+          inside: {
+            stopBy: 'end',
+            kind: 'import_statement',
+            field: 'source',
+          },
+        },
+        {
+          inside: {
+            stopBy: 'end',
+            kind: 'export_statement',
+            field: 'source',
+          },
+        },
+        {
+          inside: {
+            kind: 'string',
+            inside: {
+              kind: 'arguments',
+              inside: {
+                kind: 'call_expression',
+                has: {
+                  field: 'function',
+                  regex: '^(import|require)$',
+                },
+              },
+            },
+          },
+        },
+      ],
+    },
+  };
+  const matchModule = sgNode.findAll(matcher).map((matchNode) => {
+    return {
+      n: matchNode.text(),
+      s: matchNode.range().start.index,
+      e: matchNode.range().end.index,
+    };
+  });
+
+  for (const imp of matchModule) {
+    const { n: importPath, s: start, e: end } = imp;
+
+    if (!importPath) continue;
+
+    try {
+      const result = loadConfig(tsconfigPath);
+
+      if (result.resultType === 'failed') {
+        logger.error(result.message);
+        return;
+      }
+
+      const { absoluteBaseUrl, paths, mainFields, addMatchAll } = result;
+      const matchPath = createMatchPath(
+        absoluteBaseUrl,
+        paths,
+        mainFields,
+        addMatchAll,
+      );
```

Comment:
This will be called multiple times as it's in a for loop, consider taking it out of for loop and cache.